code/modules/onestar/os_turret.dm

```diff
@@ -0,0 +1,306 @@
+/obj/machinery/power/os_turret
+	name = "One Star turret"
+	desc = "A turret of the One Star variety."
+	icon = 'icons/obj/machines/one_star/machines.dmi'
+	icon_state = "os_gauss"
+	circuit = /obj/item/electronics/circuitboard/os_turret
+	idle_power_usage = 30
+	active_power_usage = 2500
+	density = TRUE
+	anchored = TRUE
+
+	// Targeting
+	var/should_target_players = TRUE			// TRUE targets players, FALSE targets superior animals (roaches, golems, and spiders)
+	var/range = 8								// Starts firing just out of player sight
+	var/returning_fire = FALSE					// Will attempt to fire at the nearest target when attacked and no one is in range
+
+	// Shooting
+	var/obj/item/projectile/projectile = /obj/item/projectile/bullet/gauss
+	var/number_of_shots = 5
+	var/time_between_shots = 0.5 SECONDS
+	var/list/shot_timer_ids = list()
+	var/cooldown_time = null
+
+	// Internal
+	var/emp_cooldown = 4 SECONDS
+	var/emp_timer_id
+	var/on_cooldown = FALSE
+	var/cooldown_timer_id
+
+/obj/machinery/power/os_turret/laser
+	icon_state = "os_laser"
+	circuit = /obj/item/electronics/circuitboard/os_turret/laser
+	range = 10
+	projectile = /obj/item/projectile/beam/pulse
+	number_of_shots = 3
+	time_between_shots = 0.3 SECONDS
+	cooldown_time = 2 SECONDS
+
+/obj/machinery/power/os_turret/Initialize()
+	. = ..()
+	update_icon()
+
+	if(!cooldown_time)
+		cooldown_time = time_between_shots * number_of_shots
+
+/obj/machinery/power/os_turret/Process()
+	if(stat)
+		if(prob(2))
+			do_sparks(1, TRUE, src)
+		return
+
+	if(machine_integrity <= 0)
+		stat |= BROKEN
+		return
+
+	if(!on_cooldown)
+		use_power = IDLE_POWER_USE
+	else
+		use_power = ACTIVE_POWER_USE
+		return
+
+	update_icon()
+
+	var/list/potential_targets = list()
+	var/mob/nearest_valid_target
+	var/nearest_valid_target_distance
+
+	if(should_target_players)
+		potential_targets = (GLOB.player_list & SSmobs.mob_living_by_zlevel[(get_turf(src)).z])
+	else
+		potential_targets = (GLOB.superior_animal_list & SSmobs.mob_living_by_zlevel[(get_turf(src)).z])
```

Comment:
Doesn't this object have it's own `.z`? Which is always the same as it's turf's.

Reply:
Yeah, this was a lazy copy pasta from something else. I’ll change it before merge.